mist-common/src/main/java/edu/snu/mist/common/operators/StateTransitionOperator.java

```diff
@@ -40,104 +36,126 @@
 
-    private static final Logger LOG = Logger.getLogger(StateTransitionOperator.class.getName());
-
-    // initial state
-    private final String initialState;
-
-    // current state
-    private String currState;
-
-    // final state
-    private final Set<String> finalState;
-
-    // changing state table: Map<current state, Tuple2<condition, next state>>
-    private final Map<String, Collection<Tuple2<MISTPredicate, String>>> stateTable;
-
-    /**
-     * The latest Checkpoint Timestamp.
-     */
-    private long latestCheckpointTimestamp;
-
-    @Inject
-    private StateTransitionOperator(
-        @Parameter(InitialState.class) final String initialState,
-        @Parameter(FinalState.class) final Set<String> finalState,
-        @Parameter(StateTable.class) final String stateTable,
-        final ClassLoader classLoader) throws IOException, ClassNotFoundException {
-        this(
-            initialState,
-            finalState,
-            SerializeUtils.deserializeFromString(stateTable, classLoader));
+  private static final Logger LOG = Logger.getLogger(StateTransitionOperator.class.getName());
+
+  // initial state
+  private final String initialState;
+
+  // current state
+  private String currState;
+
+  // final state
+  private final Set<String> finalState;
+
+  // changing state table: Map<current state, Tuple2<condition, next state>>
+  private final Map<String, Collection<Tuple2<MISTPredicate, String>>> stateTable;
+
+  /**
+   * The latest Checkpoint Timestamp.
+   */
+  private long latestCheckpointTimestamp;
+
+  /**
+   * The map of states for checkpointing.
+   * The key is the time of the checkpoint event, and the value is the state of this operator at that timestamp.
+   */
+  private Map<Long, Map<String, Object>> checkpointMap;
+
+  @Inject
+  private StateTransitionOperator(
+      @Parameter(InitialState.class) final String initialState,
+      @Parameter(FinalState.class) final Set<String> finalState,
+      @Parameter(StateTable.class) final String stateTable,
+      final ClassLoader classLoader) throws IOException, ClassNotFoundException {
+    this(
+        initialState,
+        finalState,
+        SerializeUtils.deserializeFromString(stateTable, classLoader));
+  }
+
+  public StateTransitionOperator(
+      final String initialState,
+      final Set<String> finalState,
+      final Map<String, Collection<Tuple2<MISTPredicate, String>>> stateTable) {
+    this.initialState = initialState;
+    this.currState = initialState;
+    this.finalState = finalState;
+    this.stateTable = stateTable;
+    this.latestCheckpointTimestamp = 0L;
+    this.checkpointMap = new HashMap<>();
+  }
+
+  // update state with input data
+  private void stateTransition(final Map<String, Object> inputData) {
+    // possible transition list in current state
+    final Collection<Tuple2<MISTPredicate, String>> transitionList = stateTable.get(currState);
```

Comment:
Guess using `Map` is better than using `Collection<Tuple2>`. Do you have any specific reason for using it?

Reply:
OK